lib/utils/patterns.ts

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2016-2017 Resin.io
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import _ = require('lodash');
+import Promise = require('bluebird');
+import form = require('resin-cli-form');
+import visuals = require('resin-cli-visuals');
+import ResinSdk = require('resin-sdk');
+import chalk from 'chalk';
+import validation = require('./validation');
+import messages = require('./messages');
+
+const resin = ResinSdk.fromSharedOptions();
+
+export function authenticate(options: {}) {
+	form.run([{
+		message: 'Email:',
+		name: 'email',
+		type: 'input',
+		validate: validation.validateEmail
+	}, {
+		message: 'Password:',
+		name: 'password',
+		type: 'password'
+	}], { override: options })
+	.then(resin.auth.login)
+	.then(resin.auth.twoFactor.isPassed)
+	.then((isTwoFactorAuthPassed: boolean) => {
+		if (isTwoFactorAuthPassed) { return; }
+
+		return form.ask({
+			message: 'Two factor auth challenge:',
+			name: 'code',
+			type: 'input'}).then(resin.auth.twoFactor.challenge)
```

Comment:
This will be fixed by prettier, but I would love to see the `.then` in the next line right away :)

Reply:
Fixed